saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/RealSyncSavedSitesRepository.kt

```diff
@@ -176,8 +189,321 @@ class RealSyncSavedSitesRepository(
+        savedSitesEntitiesDao.insert(entity)
+        savedSitesRelationsDao.insert(Relation(folderId = folder.parentId, entityId = folder.id))
+        savedSitesEntitiesDao.updateModified(folder.parentId, folder.lastModified ?: DatabaseDateFormatter.iso8601())
+        return folder
+    }
+
+    override fun replaceFolder(
+        folder: BookmarkFolder,
+        children: List<String>,
+    ) {
+        Timber.d("Sync-Bookmarks: replacing ${folder.id} with children $children")
+        savedSitesRelationsDao.replaceFolder(folder, children)
+        savedSitesEntitiesDao.update(
+            Entity(
+                entityId = folder.id,
+                title = folder.name,
+                url = "",
+                type = FOLDER,
+                lastModified = folder.lastModified ?: DatabaseDateFormatter.iso8601(),
+            ),
+        )
+    }
+
+    override fun getAllFolderContentSync(folderId: String): Pair<List<Bookmark>, List<BookmarkFolder>> {
+        val entities = savedSitesEntitiesDao.allEntitiesInFolderSync(folderId)
+        val bookmarks = mutableListOf<Bookmark>()
+        val folders = mutableListOf<BookmarkFolder>()
+        entities.forEach { entity ->
+            mapEntity(entity, folderId, bookmarks, folders)
+        }
+        return Pair(bookmarks.distinct(), folders.distinct())
+    }
+
+    private fun mapEntity(
+        entity: Entity,
+        folderId: String,
+        bookmarks: MutableList<Bookmark>,
+        folders: MutableList<BookmarkFolder>,
+    ) {
+        if (entity.type == FOLDER) {
+            val numFolders = savedSitesRelationsDao.countEntitiesInFolder(entity.entityId, FOLDER)
+            val numBookmarks = savedSitesRelationsDao.countEntitiesInFolder(entity.entityId, BOOKMARK)
+            folders.add(BookmarkFolder(entity.entityId, entity.title, folderId, numBookmarks, numFolders, entity.lastModified, entity.deletedFlag()))
+        } else {
+            bookmarks.add(
+                Bookmark(
+                    id = entity.entityId,
+                    title = entity.title,
+                    url = entity.url.orEmpty(),
+                    parentId = folderId,
+                    lastModified = entity.lastModified,
+                ),
+            )
+        }
+    }
+
+    override fun replaceBookmark(
+        bookmark: Bookmark,
+        localId: String,
+    ) {
+        // check if bookmark has moved
+        val storedBookmark = getBookmarkById(bookmark.id)
+        if (storedBookmark != null) {
+            if (storedBookmark.parentId != bookmark.parentId) {
+                // bookmark has moved to another folder
+                Timber.d("Sync-Bookmarks: ${bookmark.id} has moved from folder ${storedBookmark.parentId} to ${bookmark.parentId}")
+                savedSitesRelationsDao.deleteRelationByEntityAndFolder(bookmark.id, storedBookmark.parentId)
+                savedSitesRelationsDao.insert(Relation(folderId = bookmark.parentId, entityId = bookmark.id))
```

Comment:
As discussed, this will avoid having duplicated relations:

```suggestion
                if (savedSitesRelationsDao.relationsByFolderId(bookmark.parentId).none { it.entityId == bookmark.id }) {
                    savedSitesRelationsDao.insert(Relation(folderId = bookmark.parentId, entityId = bookmark.id))
                }
```